app/domain/authentication/jwt/verify_and_decode_token.rb

```diff
@@ -0,0 +1,62 @@
+require 'jwt'
+
+module Authentication
+  module Jwt
+
+    Err = Errors::Authentication::Jwt
+    Log = LogMessages::Authentication::Jwt
+    # Possible Errors Raised:
+    # TokenExpired, TokenDecodeFailed, TokenVerifyFailed
+
+    # This class Verifies and decodes a JWT token. It doesn't connect with the issuer
+    # of the token for verification. If a token verification is needed, a verification_options
+    # object should be provided with the JWKs & algorithms required for the verification.
+    # In addition, if token claims need to be verified they should be specified in the verification_options.
+    # For example:
+    #     @verification_options = {
+    #       algorithms: @algs,
+    #       jwks: @jwks,
+    #       verify_iss: true
+    #       iss: https://token-issuer.com/
+    #     }
+    #
+    # For more information on the verification and decode process: https://github.com/jwt/ruby-jwt
+    VerifyAndDecodeToken = CommandClass.new(
+      dependencies: {
+        jwt_decoder: JWT,
+        logger:      Rails.logger
+      },
+      inputs:       %i(token_jwt verification_options)
+    ) do
+
+      def call
+        verified_and_decoded_token
+      end
+
+      private
+
+      def verified_and_decoded_token
+        # @jwt_decoder.decode returns an array with one decoded token so we take the first object
+        @decoded_token = @jwt_decoder.decode(
+          @token_jwt,
+          nil, # the key will be taken from options[:jwks] if present
+          should_verify,
+          @verification_options
+        ).first.tap do
```

Comment:
`.first.tap` will perform the decode first and then return the object for logging?

Reply:
No. As I mention in the comment - `# @jwt_decoder.decode returns an array with one decoded token so we take the first object`

`first` is a method to get the first object in an array.
`tap` is something else - https://medium.com/aviabird/ruby-tap-that-method-90c8a801fd6a